python/pyspark/sql/connect/client/retries.py

```diff
@@ -0,0 +1,282 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import grpc
+import random
+import time
+import typing
+from typing import Optional, Callable, Generator, List, Type
+from types import TracebackType
+from pyspark.sql.connect.client.logging import logger
+
+"""
+This module contains retry system. The system is designed to be
+significantly customizable.
+
+A key aspect of retries is RetryPolicy class, describing a single policy.
+There can be more than one policy defined at the same time. Each policy
+determines which error types it can retry and how exactly.
+
+For instance, networking errors should likely be retried differently that
+remote resource being available.
+
+Given a sequence of policies, retry logic applies all of them in sequential
+order, keeping track of different policies budgets.
+"""
+
+
+class RetryPolicy:
+    """
+    Describes key aspects of RetryPolicy.
+
+    It's advised that different policies are implemented as different subclasses.
+    """
+
+    def __init__(
+        self,
+        max_retries: Optional[int] = None,
+        initial_backoff: int = 1000,
+        max_backoff: Optional[int] = None,
+        backoff_multiplier: float = 1.0,
+        jitter: int = 0,
+        min_jitter_threshold: int = 0,
+    ):
+        self.max_retries = max_retries
+        self.initial_backoff = initial_backoff
+        self.max_backoff = max_backoff
+        self.backoff_multiplier = backoff_multiplier
+        self.jitter = jitter
+        self.min_jitter_threshold = min_jitter_threshold
+
+    @property
+    def name(self):
+        return self.__class__.__name__
```

Comment:
using `self.__class__.__name__` will prevent us from registering a separate instance of the same class with different parameters, e.g. different number of retries, or different timing.